[exporter/otlp] split into its own module

codeboten opened the pull request:
Fixes #6195

bogdandrutu commented on cmd/otelcorecol/builder-config.yaml:
```diff
@@ -28,5 +28,6 @@ processors:
 
 replaces:
   - go.opentelemetry.io/collector => ../../
-  - go.opentelemetry.io/collector/semconv => ../../semconv
```
Should you not have a change around the exporter module at line 16?

codeboten replied:
Good point, updated.

bogdandrutu replied:
Same thing in the default config in the builder.

codeboten replied:
i forgot to run `make ocb`. should probably add a check for this

bogdandrutu reviewed (commented):
I expect a change in the builder

bogdandrutu commented on exporter/otlpexporter/go.mod:
```diff
@@ -0,0 +1,50 @@
+module go.opentelemetry.io/collector/exporter/otlpexporter
+
+go 1.19
```
1.18?

codeboten replied:
👍🏻

bogdandrutu commented on exporter/otlpexporter/Makefile:
```diff
@@ -0,0 +1 @@
+include ../../Makefile.Common
```
empty line please.

codeboten replied:
Done

bogdandrutu commented:
I think we should send also a PR to remove https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/doc.go since it belongs (will) to a module without anything in that parent directory. Also think what we want to do with https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter#general-information since it feels strange that it belongs to a different module.

Maybe we create a module "exporter" with all the exporters?

codeboten commented:
> I think we should send also a PR to remove https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/doc.go since it belongs (will) to a module without anything in that parent directory

I agree this makes sense once all the other exporters are split into their own modules.

> Also think what we want to do with https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter#general-information since it feels strange that it belongs to a different module.

This could likely be moved under the docs directory in something called exporter.md? Maybe there should be a components directory under docs?

> Maybe we create a module "exporter" with all the exporters?

What would the result of someone including that module be? All the exporters get pulled in? I wonder if it would be confusing for users as it may be unclear what exporters fall under that package.

bogdandrutu commented:
> What would the result of someone including that module be? All the exporters get pulled in? I wonder if it would be confusing for users as it may be unclear what exporters fall under that package.

"All the exporters get pulled in" pulled in where? All the exporters are available in the module, if you talk about the builder you need 2 lines one for module and one for import.

bogdandrutu commented:
last one about this "structure" part:
I saw somewhere in contrib that `go mod` complains if you have a module `go.opentelemetry.io/collector/exporter` (in case in the future we want to use that package and module for some shared code between exporters) and `go.opentelemetry.io/collector/exporter/otlpexporter`. Can you check that?

bogdandrutu commented:
It looks to be related to this issue https://github.com/golang/go/issues/30590, found starting from the code :)) https://cs.opensource.google/go/go/+/master:src/cmd/go/testdata/script/mod_dot.txt;l=30;bpv=1;bpt=0

codeboten commented:
Another update to this PR here:
* moved the exporter to `exporters` directory. this is more inline with opentelemetry-go: https://github.com/open-telemetry/opentelemetry-go/tree/main/exporters
* renamed `otlpexporter` to `otlpgrpcexporter` to make this more consistent with `otlphttpexporter`

Please take a look @open-telemetry/collector-approvers

dmitryax commented:
Renaming modules/packages is a significant breaking change for the API consumers. Can you please add deprecated aliases under the old paths?

dmitryax commented:
Also I'm still not convinced that we have significant enough reasons to rename `exporter` to `exporters`. Even if I like the plural more, I think we should clearly provide the reasoning behind that in this PR or an issue. Looks like the errors that @bogdandrutu posted are not relevant anymore?

dmitryax commented on exporters/otlpgrpcexporter/config.go:
```diff
@@ -12,7 +12,7 @@
 // limitations under the License.
 
-package otlpexporter // import "go.opentelemetry.io/collector/exporter/otlpexporter"
+package otlpgrpcexporter // import "go.opentelemetry.io/collector/exporters/otlpgrpcexporter"
```
Do we rename the package only but keep `otlp` as the component name?

codeboten replied:
`otlp` in the config applies to both grpc and http exporters, my opinion is that it's fine to separate config from package name, but i don't feel strongly about it. i can revert if that's preferable.

dmitryax replied:
> otlp in the config applies to both grpc and http exporters, 

otlphttpexporter uses `otlphttp` https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/otlphttpexporter/factory.go#L33

I think we should keep module names consistent with the collector component names. I would vote for keeping `otlpexporter` without renaming it. This seems like the most important one

codeboten replied:
sorry i was thinking of the receiver 😅

mx-psi replied:
I think we should keep it as `otlpexporter`. It would otherwise be the only exporter with a mismatch between the Go module name and the configuration identifier.

codeboten commented:
> Also I'm still not convinced that we have significant enough reasons to rename `exporter` to `exporters`. Even if I like the plural more, I think we should clearly provide the reasoning behind that in this PR or an issue. Probably better to have a separate PR for that. This one is just about otlp exporter. Looks like the errors that @bogdandrutu posted are not relevant anymore?

Since this is the first time we'll publish these components as separate modules, it's the time to decide on what makes the most sense. The plural does make more sense if it's going to contain more than one exporter, and as i mentioned it does align with otel-go.

That being said, I agree that deprecating/breaking the import path for users shouldn't be done lightly.

codeboten commented:
@dmitryax I opened an issue to discuss moving components to the plural subdirectory: https://github.com/open-telemetry/opentelemetry-collector/issues/6337

bogdandrutu commented:
> Looks like the errors that @bogdandrutu posted are not relevant anymore?

@dmitryax the errors I mentioned will be removed if we rename the package since, "exporter" package will be free to use.

> Also I'm still not convinced that we have significant enough reasons to rename exporter to exporters.

@dmitryax can you confirm that the errors I mentioned are ok long term?

codeboten commented:
Closing in favour of #6343
